File: Sources/LiveWall/App/WallpaperEngine.swift
import AppKit

/// Ties the pieces together: one `ScreenController` per display, a shared
/// `PowerMonitor`, and the current selection from the `Library`.
///
/// Each display gets its own source instance, so a two-monitor setup decodes
/// twice. Sharing one decoder across displays would be cheaper, but only when
/// both are visible — and the common case is that one of them is covered, where
/// separate sources let the covered one tear down independently. Worth
/// revisiting if three-plus display setups turn out to be common.
final class WallpaperEngine {

    let library = Library()
    let power = PowerMonitor()

    private var controllers: [CGDirectDisplayID: ScreenController] = [:]
    private var pendingPrepares = 0

    /// Bumped on every selection change so that an asset load finishing late
    /// can tell whether it is still the selection anyone wants.
    private var generation: UInt64 = 0

    /// Called when something the status menu displays has changed.
    var onStateChange: (() -> Void)?

    func start() {
        power.pauseOnBattery = library.pauseOnBattery
        power.onChange = { [weak self] in
            guard let self else { return }
            self.controllers.values.forEach { $0.evaluate() }
            self.onStateChange?()
        }
        power.start()

        NotificationCenter.default.addObserver(
            self, selector: #selector(screensChanged),
            name: NSApplication.didChangeScreenParametersNotification, object: nil)

        syncScreens()
        applySelection()
    }

    // MARK: - Screens

    /// `didChangeScreenParameters` fires for far more than displays being
    /// plugged in — resolution changes, wake, menu bar changes, Stage Manager.
    /// Rebuilding every source on each one meant tearing down and recreating a
    /// decoder several times a minute, which both churned CPU and grew the
    /// process's footprint steadily. Existing screens only need their geometry
    /// updated; only genuinely new screens need a source built.
    @objc private func screensChanged() {
        let added = syncScreens()
        if !added.isEmpty { applySelection(to: added) }
        onStateChange?()
    }

    @discardableResult
    private func syncScreens() -> [ScreenController] {
        var seen = Set<CGDirectDisplayID>()
        var added: [ScreenController] = []

        for screen in NSScreen.screens {
            let id = ScreenController.displayID(of: screen)
            seen.insert(id)
            if let existing = controllers[id] {
                existing.updateScreen(screen)
            } else {
                let controller = ScreenController(screen: screen, power: power)
                controller.onStateChange = { [weak self] in self?.onStateChange?() }
                controllers[id] = controller
                added.append(controller)
            }
        }

        for (id, controller) in controllers where !seen.contains(id) {
            controller.setSource(nil)
            controllers.removeValue(forKey: id)
        }

        return added
    }

    // MARK: - Selection

    func select(itemID: UUID?) {
        library.selectedID = itemID
        applySelection()
        onStateChange?()
    }

    func applySelection() {
        applySelection(to: Array(controllers.values))
    }

    private func applySelection(to targets: [ScreenController]) {
        guard !targets.isEmpty else { return }
        // Any in-flight asset load from a previous selection is now stale.
        generation &+= 1
        guard let id = library.selectedID, let item = library.item(withID: id) else {
            useShader(on: targets)
            return
        }
        useVideo(item, on: targets)
    }

    /// The procedural mode. Prefers the Core Animation gradient — it needs no
    /// Metal device at all and the compositor does the interpolation — and only
    /// uses the Metal shader when a precompiled metallib was bundled. Compiling
    /// that shader at runtime instead would cost ~97 MB, which is more than the
    /// rest of the app combined and defeats the point of the mode.
    static func makeProceduralSource() -> WallpaperSource {
        if Bundle.main.url(forResource: "wallpaper", withExtension: "metallib") != nil {
            return ShaderSource(fps: 15)
        }
        return GradientSource()
    }

    private func useShader(on targets: [ScreenController]) {
        for controller in targets {
            controller.setSource(Self.makeProceduralSource())
        }
        onStateChange?()
    }

    private func useVideo(_ item: WallpaperItem, on targets: [ScreenController]) {
        let url = library.url(for: item)
        guard FileManager.default.fileExists(atPath: url.path) else {
            Log.error("selected wallpaper missing on disk, falling back to procedural")
            library.selectedID = nil
            useShader(on: targets)
            return
        }

        let token = generation
        let fitMode = library.fitMode
        for controller in targets {
            let source = VideoSource(url: url, fps: item.fps,
                                     bitDepth: item.pixelBitDepth, fitMode: fitMode)
            pendingPrepares += 1
            source.prepare { [weak self, weak controller] ok in
                guard let self else { return }
                self.pendingPrepares -= 1

                // The selection changed while the asset was loading; this
                // source would otherwise install itself over the newer one.
                guard token == self.generation else { return }

                guard ok else {
                    Log.error("could not prepare \(item.title); falling back to procedural")
                    controller?.setSource(Self.makeProceduralSource())
                    self.onStateChange?()
                    return
                }
                controller?.setSource(source)
                self.onStateChange?()
            }
        }
    }

    // MARK: - Status

    var isAnyRendering: Bool {
        controllers.values.contains { $0.isRendering }
    }

    var screenCount: Int { controllers.count }

    /// One-line summary for the top of the status menu.
    var statusLine: String {
        if let reason = power.blockReason() {
            return "Paused — \(reason)"
        }
        if pendingPrepares > 0 { return "Loading…" }
        if controllers.isEmpty { return "No displays" }
        if isAnyRendering { return "Rendering" }
        let nearlyCovered = controllers.values.contains { $0.isNearlyCovered }
        return nearlyCovered ? "Paused — desktop nearly covered" : "Paused — desktop covered"
    }

    /// Applied to live sources in place — no reload, no decode interruption.
    func setFitMode(_ mode: FitMode) {
        library.fitMode = mode
        controllers.values.forEach { $0.setFitMode(mode) }
        onStateChange?()
    }

    func setPauseOnBattery(_ value: Bool) {
        library.pauseOnBattery = value
        power.pauseOnBattery = value
        onStateChange?()
    }
}

File: platforms/windows/src/app/WallpaperEngine.cpp
#include "app/WallpaperEngine.h"

#include <algorithm>
#include <vector>

#include "render/GradientSource.h"
#include "render/VideoSource.h"
#include "support/Log.h"
#include "support/Paths.h"
#include "support/Strings.h"

namespace livewall {
namespace {

BOOL CALLBACK collectMonitor(HMONITOR monitor, HDC, LPRECT, LPARAM parameter) {
    auto* found = reinterpret_cast<std::vector<HMONITOR>*>(parameter);
    found->push_back(monitor);
    return TRUE;
}

std::vector<HMONITOR> currentMonitors() {
    std::vector<HMONITOR> found;
    EnumDisplayMonitors(nullptr, nullptr, &collectMonitor, reinterpret_cast<LPARAM>(&found));
    return found;
}

}  // namespace

WallpaperEngine::WallpaperEngine() = default;

WallpaperEngine::~WallpaperEngine() { stop(); }

void WallpaperEngine::start(HWND window) {
    window_ = window;

    device_ = D3DDevice::acquire();
    if (!device_) {
        Log::error("no usable Direct3D 11 device — nothing can be drawn");
        return;
    }

    power_.setPauseOnBattery(library_.pauseOnBattery());
    power_.onChange = [this] {
        evaluateAll();
        if (onStateChange) onStateChange();
    };
    power_.start(window_);

    syncMonitors(/*applyToNew=*/false);
    applySelection();

    // The coverage poll. Windows can be moved and resized with no notification
    // anyone can observe, so this is the only way to learn that the desktop has
    // been covered by a window that was already open.
    SetTimer(window_, kVisibilityTimerId, MonitorController::kVisibilityPollMs, nullptr);
}

void WallpaperEngine::stop() {
    if (window_ != nullptr) {
        KillTimer(window_, kVisibilityTimerId);
        KillTimer(window_, kDeactivateTimerId);
        KillTimer(window_, kDisplayChangeTimerId);
    }
    power_.stop();
    controllers_.clear();
    device_.reset();
    window_ = nullptr;
}

void WallpaperEngine::syncMonitors(bool applyToNew) {
    if (!device_) return;

    const std::vector<HMONITOR> present = currentMonitors();

    std::vector<MonitorController*> added;
    for (const HMONITOR monitor : present) {
        const auto existing = controllers_.find(monitor);
        if (existing != controllers_.end()) {
            existing->second->refreshGeometry();
            continue;
        }

        auto controller = MonitorController::create(monitor, device_, &power_);
        if (!controller) {
            Log::error("could not build a desktop window for a monitor");
            continue;
        }
        controller->onStateChange = [this] {
            if (onStateChange) onStateChange();
        };
        MonitorController* raw = controller.get();
        controllers_.emplace(monitor, std::move(controller));
        added.push_back(raw);
    }

    // Drop controllers for monitors that are gone. HMONITOR handles are reused
    // by the OS, so identity has to come from the current enumeration rather
    // than from anything cached.
    for (auto it = controllers_.begin(); it != controllers_.end();) {
        const bool stillPresent =
            std::find(present.begin(), present.end(), it->first) != present.end();
        it = stillPresent ? std::next(it) : controllers_.erase(it);
    }

    if (applyToNew) {
        for (MonitorController* controller : added) applySelectionTo(*controller);
    }
}

void WallpaperEngine::rebuildDesktopWindows() {
    Log::info("Explorer restarted — rebuilding the desktop windows");

    // The WorkerW every window was parented into is gone, and so is the cached
    // handle for it.
    controllers_.clear();
    DesktopHost::invalidateParent();

    syncMonitors(/*applyToNew=*/false);
    applySelection();
}

std::unique_ptr<WallpaperSource> WallpaperEngine::makeProceduralSource() {
    return std::make_unique<GradientSource>(device_);
}

std::unique_ptr<WallpaperSource> WallpaperEngine::makeVideoSource(const WallpaperItem& item) {
    auto source = std::make_unique<VideoSource>(device_, library_.pathFor(item), item.fps,
                                                item.pixelBitDepth(), library_.fitMode());
    ++pendingPrepares_;
    const bool ok = source->prepare();
    --pendingPrepares_;

    if (!ok) return nullptr;
    return source;
}

void WallpaperEngine::applySelection() {
    for (auto& [monitor, controller] : controllers_) applySelectionTo(*controller);
    if (onStateChange) onStateChange();
}

void WallpaperEngine::applySelectionTo(MonitorController& controller) {
    fellBackToProcedural_ = false;

    const std::string id = library_.selectedId();
    if (id.empty()) {
        controller.setSource(makeProceduralSource());
        return;
    }

    const WallpaperItem* item = library_.item(id);
    if (item == nullptr || !paths::fileExists(library_.pathFor(*item))) {
        Log::error("selected wallpaper missing on disk, falling back to procedural");
        library_.setSelectedId({});
        fellBackToProcedural_ = true;
        controller.setSource(makeProceduralSource());
        return;
    }

    auto video = makeVideoSource(*item);
    if (!video) {
        Log::error("could not prepare " + item->title + "; falling back to procedural");
        fellBackToProcedural_ = true;
        controller.setSource(makeProceduralSource());
        return;
    }
    controller.setSource(std::move(video));
}

void WallpaperEngine::select(const std::string& itemId) {
    library_.setSelectedId(itemId);
    applySelection();
}

void WallpaperEngine::setFitMode(FitMode mode) {
    library_.setFitMode(mode);
    // Applied to live sources in place — no reload, no decode interruption.
    for (auto& [monitor, controller] : controllers_) controller->setFitMode(mode);
    if (onStateChange) onStateChange();
}

void WallpaperEngine::setPauseOnBattery(bool value) {
    library_.setPauseOnBattery(value);
    power_.setPauseOnBattery(value);
    if (onStateChange) onStateChange();
}

void WallpaperEngine::evaluateAll() {
    bool anyPending = false;
    for (auto& [monitor, controller] : controllers_) {
        controller->evaluate();
        anyPending = anyPending || controller->hasPendingDeactivate();
    }

    // A single one-shot timer for every pending teardown rather than one per
    // monitor. They all want the same delay, and a timer per display on a
    // six-monitor machine is six wake-ups where one will do.
    if (anyPending && window_ != nullptr) {
        SetTimer(window_, kDeactivateTimerId, MonitorController::kDeactivateDelayMs, nullptr);
    }
}

bool WallpaperEngine::handleMessage(UINT message, WPARAM wParam, LPARAM lParam) {
    switch (message) {
        case WM_TIMER:
            if (wParam == kVisibilityTimerId) {
                evaluateAll();
                return true;
            }
            if (wParam == kDeactivateTimerId) {
                KillTimer(window_, kDeactivateTimerId);
                for (auto& [monitor, controller] : controllers_) {
                    controller->deactivateIfStillBlocked();
                }
                return true;
            }
            if (wParam == kDisplayChangeTimerId) {
                KillTimer(window_, kDisplayChangeTimerId);
                syncMonitors(/*applyToNew=*/true);
                evaluateAll();
                if (onStateChange) onStateChange();
                return true;
            }
            return false;

        case WM_DISPLAYCHANGE:
        case WM_DPICHANGED:
            // Coalesced through a short timer. A resolution change or a monitor
            // being plugged in produces a burst of these — one per display, plus
            // more as the shell rearranges — and reacting to each would tear
            // every decoder down and rebuild it several times over.
            if (window_ != nullptr) SetTimer(window_, kDisplayChangeTimerId, 500, nullptr);
            return true;

        case WM_SETTINGCHANGE:
            // Covers a work-area change (taskbar moved or auto-hidden) as well
            // as the screen-saver settings the power monitor watches for.
            if (window_ != nullptr) SetTimer(window_, kDisplayChangeTimerId, 500, nullptr);
            return false;  // the power monitor wants this one too

        default:
            return false;
    }
}

bool WallpaperEngine::isAnyRendering() const {
    for (const auto& [monitor, controller] : controllers_) {
        if (controller->isRendering()) return true;
    }
    return false;
}

const MonitorController* WallpaperEngine::primaryController() const {
    POINT cursor{};
    GetCursorPos(&cursor);
    const HMONITOR monitor = MonitorFromPoint(cursor, MONITOR_DEFAULTTOPRIMARY);

    const auto it = controllers_.find(monitor);
    if (it != controllers_.end()) return it->second.get();
    return controllers_.empty() ? nullptr : controllers_.begin()->second.get();
}

std::string WallpaperEngine::statusLine() const {
    if (const auto reason = power_.blockReason()) return "Paused — " + *reason;
    if (pendingPrepares_ > 0) return "Loading…";
    if (controllers_.empty()) return "No displays";
    if (isAnyRendering()) {
        // Worth saying out loud: the user picked a video and is looking at a
        // gradient, and without this the menu would claim everything is fine.
        return fellBackToProcedural_ ? "Rendering — fell back to procedural" : "Rendering";
    }

    for (const auto& [monitor, controller] : controllers_) {
        if (controller->isNearlyCovered()) return "Paused — desktop nearly covered";
    }
    return "Paused — desktop covered";
}

}  // namespace livewall
